feat(api): add /user/get-session-user

ojeytonwilliams opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes: https://github.com/freeCodeCamp/freeCodeCamp/issues/50735

<!-- Feel free to add any additional description of changes below this line -->

ojeytonwilliams committed:
feat: add default user

ojeytonwilliams committed:
feat: require auth for get-session-user

ojeytonwilliams committed:
feat: reject with empty user if no username

ojeytonwilliams committed:
feat: return result: username

ojeytonwilliams committed:
feat: returns username as result

ojeytonwilliams committed:
chore: explicitly add mongodb

It's already being installed via connect-mongo, but this lets us require
it.

ojeytonwilliams committed:
feat: return joinDate

ojeytonwilliams committed:
feat: return username (or usernameDisplay)

ojeytonwilliams committed:
test: confirm return of most user properties

ojeytonwilliams committed:
test: confirm strict equality

ojeytonwilliams committed:
fix: only return expected properties

ojeytonwilliams committed:
fix: stop returning null properties

ojeytonwilliams committed:
fix: remove nulls from completedChallenges

ojeytonwilliams committed:
feat: return calendar

ojeytonwilliams committed:
fix: add isCollegeAlgebraPyCertV8

ojeytonwilliams committed:
feat: add points + isCollegeAlgebraPyCertV8

ojeytonwilliams committed:
test: delete unused properties

ojeytonwilliams committed:
fix: handle no progressTimestamp case

ojeytonwilliams committed:
feat: create getPoints helper

ojeytonwilliams committed:
fix: actually return points

ojeytonwilliams committed:
feat: return completedChallengeCount

ojeytonwilliams committed:
feat: normalize twitter handles

ojeytonwilliams committed:
fix(test): clean up after /reset-progress

Also fixed error from rebase

ojeytonwilliams committed:
feat: return userToken

ShaunSHamilton commented on api/package.json:
```diff
@@ -16,12 +16,15 @@
     "fastify-auth0-verify": "^1.0.0",
     "fastify-plugin": "^4.3.0",
+    "jsonwebtoken": "9.0.0",
+    "mongodb": "4",
```
I added `bson` instead of `mongodb`, because I thought it is smaller. Also, I wonder what coupling this would add between prisma and our package.

ojeytonwilliams replied:
I did, initially, but then I realised we already have `mongodb` as a peer dep of connect mongo :man_shrugging:

This way it's explicit and we get to use it.

> what coupling this would add between prisma and our package

None, prisma does its own thing. I can only see there being an issue if ObjectID changed and... I really hope they don't!

ShaunSHamilton commented on api/src/routes/user.test.ts:
We could keep your tests 😋

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -1,12 +1,28 @@
+import { normalizeTwitter } from '../utils/normalize';
+import { encodeUserToken } from '../utils/user-token';
+
+const removeNulls = (obj: Record<string, unknown>) =>
+  _.pickBy(obj, value => value !== null);
```
I was wondering what to do about the `null`s:
> // TODO: Handle `null` on server?

So, you think this should be handled by the server? We should not just send back `null` or `<value>`

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -90,6 +106,119 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+    },
+    async (req, res) => {
+      try {
+        // TODO: promise all to fetch both
+        const userToken = await fastify.prisma.userToken.findFirst({
+          where: { userId: req.session.user.id }
+        });
```
I decided to return a `500` with `danger` flash here. I realise this is a divergence from the old API, but we are talking about a major bug.

ShaunSHamilton replied:
Maybe we want a flash, but still to present the older of the double-user.

ojeytonwilliams replied:
This is userTokens, though. They can have more than one in principle.

```
      const tokenInfo = await UserToken.findOne({
        where: { id: userToken }
      });

      if (!tokenInfo) return res.send('User token not found');

      const { userId } = tokenInfo;

      // check if user exists for user token
      const user = await User.findOne({
        where: { id: userId }
      });
```

is how they're used right now.

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -90,6 +106,119 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+            [user.username]: {
+              ...removeNulls(publicUser),
+              completedChallenges: completedChallenges.map(removeNulls),
+              completedChallengeCount: completedChallenges.length,
```
I removed this, because the client does not even use it.

ShaunSHamilton replied:
Yes 🙈 I misspelt it as `conpletedChallengeCount` and was searching that - it is not used anywhere.

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -90,6 +106,119 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+        });
+
+        const encodedToken = userToken
+          ? encodeUserToken(userToken.id)
```
I forgot to encode the token

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -90,6 +106,119 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+
+        const user = await fastify.prisma.user.findUnique({
+          where: { id: req.session.user.id },
+          select: {
```
Did you leave anything out 😅

ojeytonwilliams replied:
No... certainly not :cold_sweat:

ShaunSHamilton commented on api/src/utils/default-user.ts:
```diff
@@ -0,0 +1,58 @@
+    showTimeLine: false
+  },
+  progressTimestamps: [],
+  // TODO: check what this is used for in api-server and if we need it
+  rand: 0,
```
No, we do not need this. It was related to our days of saving passwords

ojeytonwilliams replied:
Okay, I'll create a PR removing it from the schema.

ShaunSHamilton commented on api/src/utils/default-user.ts:
```diff
@@ -0,0 +1,58 @@
+// TODO: audit this object to find out which properties need to be updated.
+import { type Prisma } from '@prisma/client';
+
+export const defaultUser: Omit<Prisma.userCreateInput, 'email'> = {
```
Thought: We should use something like this for the seed script.

ojeytonwilliams replied:
The type checking? Could do.

ShaunSHamilton replied:
Oh. No, I just meant instead of having so much "fixture" around, we just re-use this.

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -90,6 +106,119 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+                progressTimestamps as ProgressTimestamp[] | undefined
+              ),
+              // This assertion is necessary until the database is normalized.
+              points: getPoints(
```
`points` is not a single number

ojeytonwilliams replied:
I thought that initially, but it's generated here: https://github.com/freeCodeCamp/freeCodeCamp/blob/512bcc0a8fdd9c527286e6e08990fb5d34fe4442/api-server/src/server/utils/user-stats.js#L115-L117

ojeytonwilliams replied:
It's all a confusing mess, but if you manually set `instance.points = 123` you can see that the points returned are `123`.

ShaunSHamilton commented on api/src/routes/user.ts:
```diff
@@ -90,6 +106,119 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+                progressTimestamps as ProgressTimestamp[] | undefined
+              ),
+              joinDate: new ObjectId(user.id).getTimestamp(),
+              ...normalizeTwitter(twitter),
```
I would prefer:

```suggestion
              twitter: normalizeTwitter(twitter),
```

And the function signature to match.

ojeytonwilliams replied:
The latest changes should be okay.

ShaunSHamilton reviewed (commented):
Quick thoughts based on my impl.

ojeytonwilliams commented on api/src/schemas.ts:
```diff
@@ -146,6 +146,21 @@ export const schemas = {
+  getSessionUser: {
+    response: {
+      200: Type.Object({
+        // Unfortunately, it's not possible to set a real schema for this
+        // object, since the username is dynamic and used as a key.
+        user: Type.Any(),
```
I'll see if I can come up with something better, but the fundamental problem is that we have a silly data-structure that doesn't neatly map onto JSON schemas.

Or... I'm doing it wrong. I'll see if I can do it right.

ojeytonwilliams replied:
We might be able to hack something together with `Type.String({ maxLength: 1024 })` and `Type.Record`.

Maybe. It's going to be nasty, though.

ojeytonwilliams replied:
Alrighty. A record does seem to work as we want, but it's tripping the schema security tests.

I see a few ways forward

1. Give up type safety for now
2. Don't test this schema (the regex for the keys is just `^(.*)$` which seems safe to me)
3. Fix the api-server

I'm leaning heavily towards 2. since 1. is annoying and 3. is a biggish job.

ShaunSHamilton replied:
We did discuss doing 3 many years ago, but I am also just fine with 2. Go with 2.

Although, I am super surprised this is not already an issue/feature-request on typebox.

ojeytonwilliams replied:
I did a little digging: you can, effectively, set maxLength by using `Type.String({ pattern: regexOfYourChoice })`. So, technically it can be done.

I played around a bit with setting `maxLength` explicitly, too, but that appears to only be valid for strings not patterns.

Anyways, I'll make an exception for this schema.

ojeytonwilliams committed:
refactor: change normalizeTwitter's signature

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ojeytonwilliams committed:
fix: better types for removeNulls

This keeps the keys rather than using an index signature

ojeytonwilliams committed:
feat(WIP): attempt to add full response schema

ojeytonwilliams committed:
test: prevent return of missing properties

ojeytonwilliams committed:
fix(test): ignore get-session-user's schemas

The meta schema rejects it, but the pattern should be safe.

ojeytonwilliams committed:
chore: remove log

ojeytonwilliams committed:
fix: improve getSessionUser schema

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ojeytonwilliams committed:
feat: normalize profileUI

ojeytonwilliams committed:
fix: provide type for porfolio objects

ojeytonwilliams committed:
fix: stop returning donationEmails

ojeytonwilliams committed:
feat: add shape of partiallyCompletedChallenges

ojeytonwilliams committed:
fix: add types for completedChallenges[].files

ojeytonwilliams committed:
fix: add shape of savedChallenges

ojeytonwilliams committed:
refactor: improve normalized types

ojeytonwilliams committed:
fix: set joinDate to string

ojeytonwilliams committed:
fix: correct challengeType in response

ojeytonwilliams committed:
fix: update schema with better types

If a property with an array value is missing in the db, prisma returns
an empty array. As long as the client can handle missing properties (it
can) it's fine to simply not return the property.

ojeytonwilliams committed:
feat: restore the 500 schema

ojeytonwilliams committed:
refactor: user default-user in auth

ojeytonwilliams committed:
refactor: promise all to parallelize db queries

ojeytonwilliams committed:
refactor: simplify null handling

ojeytonwilliams commented:
This one took longer than expected, even ignoring the fact that Shaun and I worked in parallel briefly.

The end result should be fairly self-explanatory.  The response schema is my current understanding of what the api-server currently returns (given the DB schema that Shaun and I put together). The rest of the code is just trying to satisfy that schema.

ojeytonwilliams committed:
test: clean up comments + add linkedin

ojeytonwilliams committed:
refactor: explain why emailVerified is true

Sembauke commented:
One of the test seemed to fail for me. It has something to do with the userToken being different from the one that is provided. 

Edit: this has been resolved on the second run!


<details><summary>Failed test locally</summary>
<p>

```bash
 FAIL  src/routes/user.test.ts (7.677 s)
  userRoutes
    Authenticated user
      /account/delete
        ✓ POST returns 200 status code with empty object (116 ms)
      /account/reset-progress
        ✓ POST returns 200 status code with empty object (38 ms)
      user/get-user-session
        ✓ GET rejects with 500 status code if the username is missing (36 ms)
        ✓ GET returns username as the result property (23 ms)
        ✓ GET returns the public user object (19 ms)
        ✕ GET returns the userToken if it exists (40 ms)
        ✓ GET returns a minimal user when all optional properties are missing (32 ms)
    Unauthenticated user
      /account/delete
        ✓ POST returns 401 status code with error message (1 ms)
      /account/reset-progress
        ✓ POST returns 401 status code with error message (2 ms)
      /get-user-session
        ✓ GET returns 401 status code with error message (1 ms)

  ● userRoutes › Authenticated user › user/get-user-session › GET returns the userToken if it exists

    expect(received).toMatchObject(expected)

    - Expected  - 1
    + Received  + 1

      Object {
    -   "userToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyVG9rZW4iOiJkdW1teS1pZCIsImlhdCI6MTY4Njg5NTM0MX0.uCLYHJoCVJYZXmXXlHKODTfJEKWti6p9HSvnt3YeB7g",
    +   "userToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyVG9rZW4iOiJkdW1teS1pZCIsImlhdCI6MTY4Njg5NTM0Mn0.7r_GYj0ydzXtN8hix1v5jczrWrxVT_B1oBSkA57fSo0",
      }

      402 |         };
      403 |
    > 404 |         expect(foobar).toMatchObject({ userToken: encodedToken });
          |                        ^
      405 |       });
      406 |       test('GET returns a minimal user when all optional properties are missing', async () => {
      407 |         // To get a minimal test user we first delete the existing one...

      at Object.<anonymous> (src/routes/user.test.ts:404:24)

A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
Test Suites: 1 failed, 12 passed, 13 total
Tests:       1 failed, 120 passed, 121 total
Snapshots:   0 total
Time:        8.435 s
Ran all test suites.
Force exiting Jest: Have you considered using `--detectOpenHandles` to detect async operations that kept running after all tests finished?
 ELIFECYCLE  Test failed. See above for more details.
sembauke@Sems-Air api % 
```

</p>
</details>

ojeytonwilliams committed:
fix: clean up after each test

ojeytonwilliams committed:
fix: delete UserToken if exists

ojeytonwilliams commented on api/src/routes/user.ts:
```diff
@@ -90,6 +108,132 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
+            username: true,
+            usernameDisplay: true,
+            website: true,
+            yearsTopContributor: true
```
Both the db and the endpoint return an array of strings. Did you spot something weird going on?

ojeytonwilliams replied:
Ah, no worries. It was quite a task to figure out exactly what gets returned, so I'm glad you've taken a look!

Sboonny commented on api/src/utils/progress.ts:
```diff
@@ -0,0 +1,22 @@
+export type ProgressTimestamp = number | { timestamp: number } | null;
+
+export const getCalendar = (
+  progressTimestamps?: ProgressTimestamp[]
+): Record<string, 1> => {
+  const calendar: Record<string, 1> = {};
+
+  progressTimestamps?.forEach(progress => {
+    if (progress === null) return;
+    if (typeof progress === 'number') {
+      calendar[Math.floor(progress / 1000)] = 1;
+    } else {
+      calendar[Math.floor(progress.timestamp / 1000)] = 1;
+    }
+  });
+
+  return calendar;
+};
+
+export const getPoints = (progressTimestamps?: ProgressTimestamp[]): number => {
+  return progressTimestamps?.length ?? 1;
+};
```
Can you give me a walkthrough of why they need to be `undefined`? will they be used without TypeScript warning about the arguments?

ojeytonwilliams replied:
Are you asking why the two functions need to be able to handle `undefined` arguments?  It's not 100% necessary, but it seemed cleaner than conditionally call the functions if `progressTimestamps` is defined.

We know (from Shaun's work defining the schema) that not all user documents have a `progressTimestamps` field, so we have to deal with the documents where it's missing.

Having said that, I think prisma will return `null`, not `undefined`, but I have to check.

> will they be used without TypeScript warning about the arguments?

I didn't quite understand this question. Could you elaborate?

Sboonny replied:
I haven't thought of the user documents and their lack of `progressTimestamps`, now this makes sense.

because of me not thinking about the user documents, I couldn't find a reason to not call it conditionally, and in attempt to understand why would it be `undefined`, I needed a walkthrough for why, thank you. 👍 

---

> I didn't quite understand this question. Could you elaborate?

I have thought that, we will need conditional call, because it didn't make sense that a call will be made to get a point or date, without the progress being made, so `progressTimestamps` will never be `undefined`.

Sboonny commented on api/src/utils/progress.test.ts:
```diff
@@ -0,0 +1,39 @@
+  });
+
+  describe('getPoints', () => {
+    it('should return 1 if there are no progressTimestamps', () => {
+      expect(getPoints(undefined)).toEqual(1);
+    });
```
This can go if it doesn't accept `undefined`.

ojeytonwilliams committed:
refactor: fix type of progressTimestamps

ojeytonwilliams committed:
fix(test): cleanup tokens after tests, not before

ojeytonwilliams committed:
fix(test): update progress utils to accept null

ojeytonwilliams commented:
I switch the labels because it's always going to have a conflict in the lock file if a weekend goes by. And those merge conflicts don't affect the reviewability (which is definitely a word).

If anyone spots a non-lock conflict, please pop the label back and I'll fix it!

ShaunSHamilton reviewed (approved):
DWISOTT

raisedadead commented:
> if a weekend goes by

If weekends go by, you have a bigger problem than label switching being inconvenient, my friend :)

Fix the conflict and merge it in, please.

ojeytonwilliams committed:
Merge branch 'main' into feat/get-session-user
